plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java

```diff
@@ -81,24 +89,77 @@ static Pom parse(IFile pomFile) throws SAXException, IOException, CoreException
     }
   }
 
-  void addDependencies(List<Library> libraries) throws CoreException {    
+  /**
+   * Select libraries whose artifacts are satisfied by the pom's dependencies.
+   */
+  public Collection<Library> resolveLibraries(Collection<Library> availableLibraries) {
+    NodeList dependenciesList = document.getElementsByTagName("dependencies");
+    if (dependenciesList.getLength() == 0) {
+      return Collections.emptyList();
+    }
+    final Element dependencies = (Element) dependenciesList.item(0);
+
+    Predicate<LibraryFile> dependencyFound = new Predicate<LibraryFile>() {
+      @Override
+      public boolean apply(LibraryFile libraryFile) {
+        Preconditions.checkNotNull(libraryFile);
+        MavenCoordinates coordinates = libraryFile.getMavenCoordinates();
+        String groupId = coordinates.getGroupId();
+        String artifactId = coordinates.getArtifactId();
+        return dependencyExists(dependencies, groupId, artifactId);
+      }
+    };
+
+    List<Library> matched = new ArrayList<>();
+    for(Library library : availableLibraries) {
+      if (Iterables.all(library.getDirectDependencies(), dependencyFound)) {
+        matched.add(library);
+      }
+    }
+    return matched;
+  }
+
+  /** Add dependencies required for the list of selected libraries. */
+  void addDependencies(List<Library> selected) throws CoreException {
+    addDependencies(selected, null);
+  }
+
+  /**
+   * Adjust dependencies required for the list of selected libraries, adding any necessary
+   * dependencies and removing any unnecessary dependencies.
+   * 
+   * @param selected the set of libraries to be included
+   * @param availableLibraries if not {@code null}, then remove any dependencies that match
```

Comment:
again, availableLibraries doesn't seem like the right name. previouslySelectedLibraries? alreadySelectedLIbraries? currentLibraries?